### PR TITLE
Feature/add habit event

### LINKED ISSUE #15: 02.02.01 Allow a user to add an optional comment to a  habit event
**User statement**
As a doer, I want a habit event to have an optional comment of up to 20 characters.

**Describe the solution you'd like**
Add a comment field to the habit event popup

**Development notes**

### PR DESCRIPTION
**Pull request by JuliePilz**
### What is it?
Added habit events! This includes the list of habit events, as well as the popup to add a new event. Skeleton code for picture selection is included, however to avoid scope creep this feature was not completed.

The following files were added:
Classes:
-HabitEvent
-HabitEventList
-AddHabitEventDialog
-HabitEventItemAdapter
-HabitEventListFragment

Layouts:
add_habit_event_dialog
fragment_habit_event_list
habit_event_row

Closes #15 and #14

### CODE REVIEW

File: app/src/main/java/com/example/habitsmasher/HabitEventList.java
Comment (JacobNguyen71) [inline app/src/main/java/com/example/habitsmasher/HabitEventList.java:29]: noticed this method is not used so I'm not sure if we'll need it for later but if not I'd remove it.

File: app/src/main/java/com/example/habitsmasher/HabitEventList.java
Comment (JacobNguyen71) [inline app/src/main/java/com/example/habitsmasher/HabitEventList.java:39]: This method had the same name as the comment I left above, if both are meant to stay, I think renaming them would be a good idea so they are not as confusing

File: app/src/main/java/com/example/habitsmasher/ui/history/HabitEventItemAdapter.java
Comment (JacobNguyen71) [inline app/src/main/java/com/example/habitsmasher/ui/history/HabitEventItemAdapter.java:91]: unused code, make sure to delete if not needed for later

File: app/src/androidTest/java/com/example/habitsmasher/MainActivityTest.java
Comment (rudy-patel) [inline app/src/androidTest/java/com/example/habitsmasher/MainActivityTest.java:116]: Instead of newing it up in each test, can we move this to be a field that gets set in the `setUp` method instead?

File: app/src/main/java/com/example/habitsmasher/Habit.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/Habit.java:86]: minor: typo in `habbitEvents`

File: app/src/main/java/com/example/habitsmasher/HabitEvent.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/HabitEvent.java:18]: Since we are transitioning to `UUID`, when do we want to update the `Habit` class to use this as well?

File: app/src/main/java/com/example/habitsmasher/HabitEvent.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/HabitEvent.java:38]: The `this` keyword isn't needed here (or in **any** of the getters/setters in this class). Since our fields start with an `_`, we don't need to differentiate the variable context.

If we named our fields the same as our local variables, then we would do:
`this._startDate = startDate`
But, since they are named **differently**, we can do:
`_startDate = startDate`

The `this` keyword is already **implied**.

File: app/src/main/java/com/example/habitsmasher/HabitEventList.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/HabitEventList.java:39]: I think they are both useful, they are the same name, but **different parameters**. This gives us the flexibility to either provide the `HabitEvent` object to the add method or supply each field separately. As long as we test both, they can both stay in my opinion.

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/AddHabitDialog.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/dashboard/AddHabitDialog.java:23]: We added this import, but it's not used in this class anywhere (no other diffs relating to `HabitEventList` in this file)

File: app/src/main/java/com/example/habitsmasher/ui/history/AddHabitEventDialog.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/history/AddHabitEventDialog.java:190]: I would prefer this to be **extracted** to its own class, `HabitEventValidator`, and see some specific unit tests for all input combinations. You can take "inspiration" from the `HabiValidator`, and also its associated tests.

File: app/src/main/java/com/example/habitsmasher/ui/history/AddHabitEventDialog.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/history/AddHabitEventDialog.java:158]: Also, a lot of the **date** stuff has been extracted out by @JacobNguyen71 as well. I think it would be good to touch base with him, or create a task to extract the date picker stuff to use the new extracted class **later** (still this sprint).

File: app/src/main/java/com/example/habitsmasher/ui/history/HabitEventItemAdapter.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/history/HabitEventItemAdapter.java:35]: Since we now have **two similar adapters**, would it make sense to introduce an **interface** that is shared between the two that extracts the commonalities between the two (context, snapshots, date format, etc.)

File: app/src/main/java/com/example/habitsmasher/ui/history/HabitEventListFragment.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/history/HabitEventListFragment.java:100]: Let's at least add a **log message** here so the body isn't empty, allowing us to step through the code while debugging in the future.

File: app/src/main/java/com/example/habitsmasher/ui/history/HabitEventListFragment.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/history/HabitEventListFragment.java:44]: Same thing here, since we have two "ListFragments" now, can we extract away any commonalities between the two lists to an interface? Things like the swipe functionality are shared between the two.

File: app/src/main/res/layout/add_habit_event_dialog.xml
Comment (rudy-patel) [inline app/src/main/res/layout/add_habit_event_dialog.xml:150]: Instead of **text**, shouldn't we be setting `android:hint` instead?

File: app/src/main/java/com/example/habitsmasher/HabitEventList.java
Comment (JuliePilz) [reply app/src/main/java/com/example/habitsmasher/HabitEventList.java:39]: I would agree with Rudy. It's the same function however with different parameters, so I don't think a different name would make a lot of sense

File: app/src/main/java/com/example/habitsmasher/ui/history/HabitEventItemAdapter.java
Comment (JuliePilz) [reply app/src/main/java/com/example/habitsmasher/ui/history/HabitEventItemAdapter.java:91]: Needed later :)

File: app/src/main/java/com/example/habitsmasher/HabitEvent.java
Comment (JuliePilz) [reply app/src/main/java/com/example/habitsmasher/HabitEvent.java:18]: Up to you, I can do it on this PR for consistency if you want

File: app/src/main/java/com/example/habitsmasher/ui/history/AddHabitEventDialog.java
Comment (JuliePilz) [reply app/src/main/java/com/example/habitsmasher/ui/history/AddHabitEventDialog.java:162]: Two avoid blocking, I'll make a task to do it once that has been merged into develop

File: app/src/main/java/com/example/habitsmasher/ui/history/HabitEventItemAdapter.java
Comment (JuliePilz) [reply app/src/main/java/com/example/habitsmasher/ui/history/HabitEventItemAdapter.java:35]: Hmmm I see where you are coming from, but I think there's actually too many differences between them to support that at this time. Only context and date format are the same, snapshots actually have a different type between the two (HabitEvent vs Habit). I definitely think it's worth looking into in the future, but it may need further discussion so perhaps in a later issue?

File: app/src/main/java/com/example/habitsmasher/ui/history/HabitEventListFragment.java
Comment (JuliePilz) [reply app/src/main/java/com/example/habitsmasher/ui/history/HabitEventListFragment.java:44]: Similarly to the other comment, although I agree that this is a good idea, it will need further investigation. Even with things like the swipe functionality that is shared, they're doing slightly different things (one removes a habit event, the other a habit). Some reworking and talk will be needed which I think should be done separately of this PR

File: app/src/main/java/com/example/habitsmasher/HabitEvent.java
Comment (jaskim9824) [reply app/src/main/java/com/example/habitsmasher/HabitEvent.java:18]: We should probably address this in the Database Entity PR, since we'll have Habit, HabitEvents etc. extend from it.

File: .idea/misc.xml
Comment (jaskim9824) [inline .idea/misc.xml:8]: These shouldn't be here, as this path is local to your machine.

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitViewFragment.java
Comment (jaskim9824) [inline app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitViewFragment.java:32]: We should pass in the user from the HabitListFragment rather than creating a new one here.

File: app/src/main/java/com/example/habitsmasher/HabitEvent.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/HabitEvent.java:18]: The database entity PR is already up for review I think. That story doesn't change any of the existing IDs. If this story is changing the IDs, I think the change should **propagate** to the other data stored in the db. This PR is introducing a new way to do IDs, so it should also **modify** the others to make it consistent. Don't want to **block** this PR, so make sure to create a task for this.
- [ ] create a task to change all IDs to UUID

File: app/src/main/java/com/example/habitsmasher/HabitEvent.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/HabitEvent.java:57]: Same thing here, the `this` keyword is not necessary. It's implied.

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitViewFragment.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitViewFragment.java:32]: I agree ^

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitValidator.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitValidator.java:62]: Why is this necessary?

File: app/src/main/java/com/example/habitsmasher/ui/history/AddHabitEventDialog.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/ui/history/AddHabitEventDialog.java:162]: Cool, I'd just make the task now and not wait until this is merged (because we will both likely forget)

File: app/src/main/java/com/example/habitsmasher/ui/history/HabitEventItemAdapter.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/ui/history/HabitEventItemAdapter.java:35]: Yup, I was never implying that we need to do it **here and now**. Just starting up the discussion and asking some clarifying questions. I think there are definitely some similarities that can be extracted away, we basically have two very similar adapters. The type differences can be solved easily using Java Generics, and **interfaces** don't provide implementations, just the declarations of what needs to be there.

File: app/src/main/java/com/example/habitsmasher/ui/history/HabitEventListFragment.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/ui/history/HabitEventListFragment.java:44]: Similar to my reply to your comment, I was just starting up a discussion here. 

I think there are clear similarities, and we now have two very similar list fragments and associated adapters. Again, since interfaces don't actually define the functionality, it doesn't matter **what** the swipe does or if it's different, we just need **some** swipe function in there, and the details can be defined later.

Essentially, this enforces OOP and builds a contract for any future classes we build, serving as a reminder that we need to implement a swipe function (for example).

File: app/src/main/java/com/example/habitsmasher/ui/history/HabitEventValidator.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/history/HabitEventValidator.java:13]: Nice!

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitValidator.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitValidator.java:62]: Nevermind, I see you use it in your new Validator

File: .idea/misc.xml
Comment (rudy-patel) [reply .idea/misc.xml:8]: - [ ] remove local path

**Review by rudy-patel — CHANGES_REQUESTED** [review]
Some good discussion in the comments, let make sure to create tasks and stuff for any of the discussion points so we can circle back to them later. Other than that, the **tasks** that I've left are things that should be modified for **this** PR. I'd also like to see 1 (or preferably more) UI tests to prove this functionality works on the front-end! After that, I think it's good to go.
- [ ] create some UI tests

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitViewFragment.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitViewFragment.java:32]: - [ ] pass in user from `HabitListFragment`

File: app/src/main/java/com/example/habitsmasher/ui/history/HabitEventListFragment.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/history/HabitEventListFragment.java:101]: Nice!

File: app/src/test/java/com/example/habitsmasher/ui/history/HabitEventListTest.java
Comment (rudy-patel) [inline app/src/test/java/com/example/habitsmasher/ui/history/HabitEventListTest.java:16]: Good to see this under test!

**Review by rudy-patel — DISMISSED** [review]
Good to see all the tests added and all the small tasks completed! 

The only obvious thing I see missing is adding your new test classes to the **test suite**. You can do that now or in your next PR, doesn't matter to me but the earlier the better.

Approved!

**Review by jaskim9824 — APPROVED** [review]

**Review by jaskim9824 — APPROVED** [review]

**Review by rudy-patel — APPROVED** [review]

**Pull request merged**
